Lib/typing.py

```diff
@@ -344,20 +344,23 @@ def _flatten_literal_params(parameters):
 
 
 _cleanups = []
+_caches = { }
 
 
 def _tp_cache(func=None, /, *, typed=False):
     """Internal wrapper caching __getitem__ of generic types with a fallback to
     original function for non-hashable arguments.
     """
     def decorator(func):
-        cached = functools.lru_cache(typed=typed)(func)
-        _cleanups.append(cached.cache_clear)
+        cache = functools.lru_cache(typed=typed)(func)
+        _caches[func] = cache
+        _cleanups.append(cache.cache_clear)
+        del cache
```

Comment:
Can we add a short comment here about why we're taking this approach (maybe linking back to the GitHub issue)?

Reply:
Done just now.